### PR TITLE
SEP support

### PR DESCRIPTION
**Pull request by worksofliam**
### Changes

* [x] Add new attribute for batch debugging (#1823)
* [x] Add new SEP port config for connections (#1893)
* [x] Pass in port and SEP port when starting debug service (#1894)
* [x] Change Java versions based on installed Debug Service (#1841)
* [x] Support starting batch and SEP debug (#1824)
   * [x] Code written to execute the SEP debug command
   * [x] UI/Commands updated/created to launch batch/SEP debug from active editor and object browser
* [x] https://github.com/codefori/vscode-ibmi/issues/1932
* [x] Ensure debug extension is activated before trying to launch SEP
* [x] Merged in #1895 (this PR depends on it)

### How to test this PR

1. [x] Ensure using Debug Service version 1 still works (✅ by @worksofliam)
2. [x] Show error if SEP is not supported (✅ by @worksofliam)
3. [x] Can launch batch from active editor (✅ by @worksofliam)
3. [x] Can launch batch from Object Browser (✅ by @mkwan01)
4. [x] Can launch SEP from active editor (✅ by @mkwan01)
5. [x] Can launch SEP from Object Browser (✅ by @mkwan01)

### Checklist

* [x] have tested my change
* [ ] have created one or more test cases
* [ ] updated relevant documentation
* [ ] Remove any/all `console.log`s I added
* [ ] have added myself to the contributors' list in [CONTRIBUTING.md](https://github.com/codefori/vscode-ibmi/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: src/api/debug/server.ts [hunk 1]
```diff
@@ -4,15 +4,55 @@ import { window } from "vscode";
...
+    java: `8`
+  }
+}
+
+export async function startup(instance: Instance){
+  const connection = instance.getConnection()!;
+  const content = instance.getContent()!;
+  
   const host = connection.currentHost;
+  const details = await getDebugServiceDetails(content);
```
Comment (edmundreinhardt) [inline src/api/debug/server.ts:59]: Not that if the details file exists but does not have the `java` entry, we will get bad behaviour with `undefined`

Comment (worksofliam) [reply src/api/debug/server.ts:59]: The API has a default return value if the file does not exist.

**Comment by worksofliam** [comment]
Some screenshots:

image

image

image

**Comment by mkwan01** [comment]
I just tested a driver with all the changes as of 03-26 (code-for-ibmi-2.8.5-dev.0.vsix). I found the following problems in the certificate generation area:
1.  The certificate is still generated under /QIBM/ProdData/IBMiDebugService/bin/certs. We are supposed to store the certificates under the following new path:
/QIBM/UserData/IBMiDebugService/certs

The path /QIBM/ProdData/IBMiDebugService/bin is only writable for a user with *QSECOFR authority, while the path /QIBM/UserData/IBMiDebugService is writable by any user.

2. Only the server certificate (the .pfx file) is generated under the current certificate path /QIBM/ProdData/IBMiDebugService/bin/certs. The client certificate (the .crt file) is not generated. 

3. The client certificate is not downloaded to the vscode client. This problem may be linked with problem #2.

4. The environment variable DEBUG_CA_PATH is not set. This problem may be linked with problem #2 and #3.

**Comment by worksofliam** [comment]
@mkwan01 

> The certificate is still generated under /QIBM/ProdData/IBMiDebugService/bin/certs.

This is because you were still using the old path in some fashion. We don't just switch for no reason. Since the old certs directory still existed (likely), then we can't switch it automatically to the new directory safely. You can still change it manually in the settings. 

> The client certificate (the .crt file) is not generated.

We no longer keep the certificate around. We generate the `crt` when the server is running for each client to download. You can see this logic here: https://github.com/codefori/vscode-ibmi/blob/master/src/api/debug/certificates.ts#L128

If your client certificate fails to download, please share the Code for IBM i output channel with me after you see it fail.

**Comment by mkwan01** [comment]
@worksofliam I deleted the "/QIBM/ProdData/IBMiDebugService/bin/certs" folder and select "Setup certificates > Generate certificates" from the "Set up the debug service on IBM i" welcome page again. The .pfx file is still generated under the /QIBM/ProdData/IBMiDebugService/bin/certs.

Here is output of the "Generate certificates" task:
```
/home/MKWAN: ls -p /QIBM/ProdData/IBMiDebugService/bin/certs/debug_service.pfx
{
    "code": 2,
    "signal": null,
    "stdout": "",
    "stderr": "ls: cannot access '/QIBM/ProdData/IBMiDebugService/bin/certs/debug_service.pfx': No such file or directory"
}

/home/MKWAN: mkdir -p /QIBM/ProdData/IBMiDebugService/bin/certs
{
    "code": 0,
    "signal": null,
    "stdout": "",
    "stderr": ""
}

/QIBM/ProdData/IBMiDebugService/bin/certs: openssl genrsa -out debug_service.key 2048 && openssl req -new -key debug_service.key -out debug_service.csr -subj '/CN=p9adt02.rch.stglabs.ibm.com' && openssl x509 -req -in debug_service.csr -signkey debug_service.key -out debug_service.crt -days 1095 -sha256 -req -extfile <(printf "subjectAltName=DNS:p9adt02.rch.stglabs.ibm.com,IP:9.5.160.191") && openssl pkcs12 -export -out debug_service.pfx -inkey debug_service.key -in debug_service.crt -password pass:p9adt02.rch.stglabs.ibm.com && rm debug_service.key debug_service.csr debug_service.crt && chmod 444 debug_service.pfx
{
    "code": 0,
    "signal": null,
    "stdout": "",
    "stderr": "Certificate request self-signature ok\nsubject=CN = p9adt02.rch.stglabs.ibm.com"
}

```

**Comment by mkwan01** [comment]
I found a problem with the client certificate downloading. On some IBM i machines, debug can only use TLSv1.2 for communication. This also causes an issue with the "openssl s_client" command to download the client certificate.

Here is the output on these machines:
```
> openssl s_client -connect localhost:8005 -showcerts                          
  CONNECTED(00000004)                                                          
  00000001:error:0A000411:SSL routines:ssl3_read_bytes:sslv3 alert no certificate:ssl/record/rec_layer_s3.c:1586:SSL alert number 41                        
  ---                                                                          
  no peer certificate available                                                
  ---                                                                          
  No client certificate CA names sent  
  ---
```
The solution is to detect the error condition (the words in BOLD are consistent) and try to add a -tls1_2 option if the previous command fails.

Here is the output of the new command:
 ```
> openssl s_client -connect localhost:8005 -tls1_2 -showcerts < /dev/null 2> 
              /dev/null | openssl x509 -outform PEM                           
   -----BEGIN CERTIFICATE-----                                                
   MIIDKTCCAhGgAwIBAgIUHIPrzkdl32iAe0z4LKwV7KjJV/kwDQYJKoZIhvcNAQEL           
   BQAwJjEkMCIGA1UEAwwbcDlhZHQwMi5yY2guc3RnbGFicy5pYm0uY29tMB4XDTI0           
   MDMyNzEwNTI0MloXDTI3MDMyNzEwNTI0MlowJjEkMCIGA1UEAwwbcDlhZHQwMi5y           
   Y2guc3RnbGFicy5pYm0uY29tMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKC           
   AQEAuA5ThO95y2+hrBNYzjtEuFEFKrJS9jdUJqWkJgjD/JiK6zadBbmgCh60cKQo
```

**Comment by sebjulliand** [comment]
> I found a problem with the client certificate downloading. On some IBM i machines, debug can only use TLSv1.2 for communication. This also causes an issue with the "openssl s_client" command to download the client certificate.

@worksofliam, here is the command that will allow to get the certificate from the keystore itself. That will avoid the issue found by @mkwan01 .

Replace
```typescript
`openssl s_client -connect localhost:${connection.config?.debugPort} -showcerts < /dev/null 2> /dev/null | openssl x509 -outform PEM`
```

By
```typescript
`openssl pkcs12 -in ${getRemoteServerCertificatePath(connection)} -passin pass:${connection.currentHost} -info -nokeys -clcerts 2>/dev/null | openssl x509 -outform PEM`
```

In the `downloadClientCert` method in `certificates.ts`.

**Pull request merged**